src/borg/archive.py

```diff
@@ -965,15 +962,15 @@ def process_file(self, path, st, cache, ignore_inode=False):
                     # block or char device will change without its mtime/size/inode changing.
                     cache.memorize_file(path_hash, st, [c.id for c in item.chunks])
                 status = status or 'M'  # regular file, modified (if not 'A' already)
+            self.stats.nfiles += 1
         item.update(self.stat_attrs(st, path))
         item.get_size(memorize=True)
         if is_special_file:
             # we processed a special file like a regular file. reflect that in mode,
             # so it can be extracted / accessed in FUSE mount like a regular file:
             item.mode = stat.S_IFREG | stat.S_IMODE(item.mode)
-        self.stats.nfiles += 1
         self.add_item(item)
-        if hardlinked and hardlink_master:
+        if hardlink_master:
             # Add the hard link reference *after* the file has been added to the archive.
             self.hard_links[st.st_ino, st.st_dev] = safe_path
```

Comment:
hmm, thinking about whether this "*after*" is important.
or whether it could be also done in line 922 else-block.